server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -219,83 +218,148 @@ public void stop()
-                              "SELECT payload FROM %s WHERE dataSource = :dataSource",
-                              getSegmentsTable()
-                          )
-                      )
-                      .setFetchSize(connector.getStreamingFetchSize())
-                      .bind("dataSource", dataSource)
-                      .map(ByteArrayMapper.FIRST)
-                      .iterator(),
-                  payload -> {
-                    try {
-                      return jsonMapper.readValue(payload, DataSegment.class);
-                    }
-                    catch (IOException e) {
-                      throw new RuntimeException(e);
-                    }
+    return connector.inReadOnlyTransaction(
+        (handle, status) -> VersionedIntervalTimeline.forSegments(
+            Iterators.transform(
+                handle
+                    .createQuery(
+                        StringUtils.format(
+                            "SELECT payload FROM %1$s WHERE dataSource = :dataSource AND start < :end AND %2$send%2$s > :start",
+                            getSegmentsTable(), connector.getQuoteString()
+                        )
+                    )
+                    .setFetchSize(connector.getStreamingFetchSize())
+                    .bind("dataSource", dataSource)
+                    .bind("start", interval.getStart().toString())
+                    .bind("end", interval.getEnd().toString())
+                    .map(ByteArrayMapper.FIRST)
+                    .iterator(),
+                payload -> {
+                  try {
+                    return jsonMapper.readValue(payload, DataSegment.class);
                   }
-              )
-
-          )
-      );
-
-      final List<DataSegment> segments = new ArrayList<>();
-      List<TimelineObjectHolder<String, DataSegment>> timelineObjectHolders = segmentTimeline.lookup(
-          Intervals.of("0000-01-01/3000-01-01")
-      );
-      for (TimelineObjectHolder<String, DataSegment> objectHolder : timelineObjectHolders) {
-        for (PartitionChunk<DataSegment> partitionChunk : objectHolder.getObject()) {
-          segments.add(partitionChunk.getObject());
-        }
-      }
-
-      if (segments.isEmpty()) {
-        log.warn("No segments found in the database!");
-        return false;
-      }
-
-      dbi.withHandle(
-          new HandleCallback<Void>()
-          {
-            @Override
-            public Void withHandle(Handle handle)
-            {
-              Batch batch = handle.createBatch();
+                  catch (IOException e) {
+                    throw new RuntimeException(e);
+                  }
+                }
+            )
+        )
+    );
+  }
 
-              for (DataSegment segment : segments) {
-                batch.add(
+  private VersionedIntervalTimeline<String, DataSegment> getVersionedIntervalTimeline(final String dataSource, final Collection<String> segmentIds)
```

Comment:
Suggest `buildVersionedIntervalTimeline` instead, with some method javadoc that explains that it builds a new timeline containing only the provided segments